### PR TITLE
Use of alias instead of IP address or hostname

### PR DESCRIPTION
**Pull request by kwazel**
Some people will have a network where TP-link devices like power plugs do not have a static ip-address and do not have the luxury of a hostname. In this case eacht TP-link device has a device name, or alias. This can be used to identify the smart plugs and bulbs uniquely. Using the parameter
--alias
instead of --ip or --hostname, the cli will attempt to discover a device with this alias (case insensitive). Upon finding such a device, it will use this device, e.g. `cli.py --alias kalle --plug off` will switch a smart plug with the alias "Kalle" or "kalle" off.
`cli.py --alias "kalle" --plug on` will switch the device on again. A little bit confusing will be
`cli.py --alias "kalle" alias "kalle anka"` which will change the device name, alias, of a device with the alias "kalle". The new alias will be "kalle anka". When undesired, this can be circumvented by:

- changing `--ip`, `--hostname` or `--alias` into `--byip`, `--byhostname` or `--byalias`
- changing `alias` as parameter into e.g. `device`
- proper documentation

### CODE REVIEW

**Comment by kwazel** [comment]
I have requested this change before, but back then (April) I did not find time to actually implement the proposed changes (I was actually fighting with the unit tests, which I am now able to run but do not cover the cli very well).

File: pyHS100/cli.py [hunk 1]
```diff
@@ -41,11 +44,17 @@ def cli(ctx, ip, host, debug, bulb, plug):
     if ip is not None and host is None:
         host = ip
 
+    if alias is not None and host is None:
+        click.echo("Alias is given, using discovery to find host %s" %
+                   alias)
+        host = find_host_from_alias(alias=alias)
+        print("Found hostname is {}".format(host))
+
+
```
Comment (rytilahti) [inline pyHS100/cli.py]: We do not have hound-ci active at the moment (which i hope we'll get enabled soonish) to notify about code-style issues, but here is an extra newline.

File: pyHS100/cli.py [hunk 2]
```diff
@@ -72,14 +81,28 @@ def discover(ctx, timeout, discover_only):
     click.echo("Discovering devices for %s seconds" % timeout)
     found_devs = Discover.discover(timeout=timeout).items()
     if not discover_only:
-        for ip, dev in found_devs:
+        for _, dev in found_devs:
```
Comment (rytilahti) [inline pyHS100/cli.py]: I think it's better to leave it as it was, as it makes it easy for a reader to see what it is behind the variable. So either revert this change, or use `found_devs.items()` and drop it altogether.

File: pyHS100/cli.py [hunk 3]
```diff
@@ -72,14 +81,28 @@ def discover(ctx, timeout, discover_only):
...
     return found_devs
 
 
+def find_host_from_alias(alias, timeout=1, attempts=3):
+    """Discover a device identified by its alias"""
+    host = None
+    click.echo("Trying to discover %s using %s attempts of %s seconds" %
+               (alias, attempts, timeout))
+    for attempt in range(1, attempts):
+        print("Attempt %s of %s" % (attempt, attempts))
```
Comment (rytilahti) [inline pyHS100/cli.py]: Use `click.echo` instead of `print`.

File: pyHS100/cli.py [hunk 4]
```diff
@@ -72,14 +81,28 @@ def discover(ctx, timeout, discover_only):
...
+        found_devs = Discover.discover(timeout=timeout).items()
+        for _, dev in found_devs:
```
Comment (rytilahti) [inline pyHS100/cli.py]: See my above note about anonymous variables.

File: pyHS100/cli.py [hunk 5]
```diff
@@ -72,14 +81,28 @@ def discover(ctx, timeout, discover_only):
...
+            if dev.alias.lower() == alias.lower():
+                host = dev.host
+                return host
+    return None
+
```
Comment (rytilahti) [inline pyHS100/cli.py:108]: This should have two empty lines (between functions as per pep8).

**Review by rytilahti — CHANGES_REQUESTED** [review]
Looks good to me, thanks for the contribution! I'm unsure if there's a need to have a envvar for this, but that's fine. I added a couple of very small, easy-to-fix issues that should be changed before merging this.

File: pyHS100/cli.py (hunk 4 above)
Comment (kwazel) [reply pyHS100/cli.py]: I see your point, I was just following Pylints advice.

File: pyHS100/cli.py [hunk 6]
```diff
@@ -41,11 +44,16 @@ def cli(ctx, ip, host, debug, bulb, plug):
     if ip is not None and host is None:
         host = ip
 
+    if alias is not None and host is None:
+        click.echo("Alias is given, using discovery to find host %s" %
+                   alias)
+        host = find_host_from_alias(alias=alias)
+        click.echo("Found hostname is {}".format(host))
```
Comment (rytilahti) [inline pyHS100/cli.py]: Could you add a check if the returned host is `None` and return if so, I'd expect that when using `--alias` it should bail out if it cannot discover the device instead of continuing. Otherwise this looks good to be merged :-)

**Review by rytilahti — APPROVED** [review]

**Comment by rytilahti** [comment]
Great, thanks! :tada:
